File: ruby/lib/dado.rb
require_relative 'diario'
require 'singleton'

module Civitas
    class Dado
        include Singleton
        @@SalidaCarcel = 5
        @@random = nil
        @@ultimoResultado = 1 #int
        @@debug = false #bool

        def initialize 
        end

        def tirar #int
            if (!@@debug)
                @@ultimoResultado = rand(1..6).floor
            end

            return @@ultimoResultado
        end

        def salgoDeLaCarcel
            tirar()
            if (@@ultimoResultado == @@SalidaCarcel)
                return true
            else
                return false
            end
        end

        def quienEmpieza(n)
            ran = rand(0...n)
            return ran
        end

        def setDebug(d)
            @@debug = d
            if (d)
                cadena = "El modo debug ha sido activado"
            else
                cadena = "El modo debug ha sido desactivado"
            end
            Diario.instance.ocurreEvento(cadena)
        end

        def ultimoResultado
            return @@ultimoResultado
        end

        def toString
            string="Salida carcel: "+@@SalidaCarcel.to_s
            string+=" Ultimo resultado: "+@@ultimoResultado.to_s
            string+=" Debug: "+@@debug.to_s
            return string
        end

    end
end
#Dado.instance para obtener la instancia

File: java/src/civitas/Dado.java
package civitas;

import java.util.Random;

public class Dado {
    private Random random;
    private int ultimoResultado;
    private boolean debug;

    private static final Dado instance = new Dado();
    private static int SalidaCarcel = 5;

    private Dado() {
        ultimoResultado = 0;
        debug = false;
        random = new Random();
    }

    public static Dado getInstance() {
        return instance;
    }

    private int getRandomNumberInRange(int min, int max) {

        if (min >= max) {
            throw new IllegalArgumentException("max must be greater than min");
        }

        random = new Random();
        return random.nextInt((max - min) + 1) + min;
    }

    int tirar() {
        if (debug) {
            ultimoResultado = 1;
            return 1;
        } else {
            int valor = getRandomNumberInRange(1, 6);
            ultimoResultado = valor;
            return valor;
        }
    }

    boolean salgoDeLaCarcel() {
        return (tirar() >= SalidaCarcel);
    }

    int quienEmpieza(int n) {
        return getRandomNumberInRange(0, n - 1);
    }

    public void setDebug(boolean d) {
        debug = d;
        String e = "Dado--Modo debug: ";
        if (d)
            e.concat("activado");
        else
            e.concat("desactivado");
        Diario.getInstance().ocurreEvento(e);
    }

    int getUltimoResultado() {
        return ultimoResultado;
    }

}